### PR TITLE
chore(deps): update typescript

### PR DESCRIPTION
**Pull request by alexander-akait**
Update typescript + plus:
1. Improve types
2. [WIP] we need update deps in our tooling

### <samp>🤖 Generated by Copilot at 8e0f742</samp>

This pull request updates the `typescript` dependency in `package.json` and enhances the TypeScript integration and compatibility for webpack.

### <samp>🤖 Generated by Copilot at 8e0f742</samp>

* Update the `typescript` dependency to 5.0.4 for new features and bug fixes ([link](https://github.com/webpack/webpack/pull/16957/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L103-R103))

### CODE REVIEW

File: lib/serialization/ObjectMiddleware.js [hunk 1]
```diff
@@ -46,11 +46,13 @@ Technically any value can be used.
 /**
  * @typedef {Object} ObjectSerializerContext
  * @property {function(any): void} write
+ * @property {function(any): void} setCircularReference
  */
 
 /**
  * @typedef {Object} ObjectDeserializerContext
  * @property {function(): any} read
+ * @property {function(any): void} setCircularReference
```
Comment (alexander-akait) [inline lib/serialization/ObjectMiddleware.js:55]: The funny thing is that we use this in some methods, but it was not in the types, and for some reason typescript doesn't catch it

Comment (TheLarkInn) [reply lib/serialization/ObjectMiddleware.js:55]: cc @danielrosenwasser

**Comment by TheLarkInn** [comment]
cc @jakebailey @danielrosenwasser going to tag you both here in case we catch any bugs that need squashing 🤣🤣🤣

**Comment by jakebailey** [comment]
webpack is a part of typescript's user tests, so I would certainly hope there's nothing unexpected! https://github.com/microsoft/typescript-error-deltas/tree/main/userTests/webpack

File: lib/util/LazySet.js [hunk 2]
```diff
@@ -187,12 +187,20 @@ class LazySet {
...
+	 */
 	serialize({ write }) {
 		if (this._needMerge) this._merge();
 		write(this._set.size);
 		for (const item of this._set) write(item);
 	}
 
+	/**
+	 * @template T
+	 * @param {import("../serialization/ObjectMiddleware").ObjectDeserializerContext} context context
```
Comment (alexander-akait) [inline lib/util/LazySet.js:201]: @jakebailey We have a bug/problem here if you rewrite this to (at top):
```
/** @typedef {import("../serialization/ObjectMiddleware").ObjectSerializerContext} ObjectSerializerContext */
/** @typedef {import("../serialization/ObjectMiddleware").ObjectDeserializerContext} ObjectSerializerContext */
```

`LazySet` starts acting weird (one from many erros):
```
lib/sharing/resolveMatchedConfigs.js:88:42 - error TS2345: Argument of type 'LazySet<string>' is not assignable to parameter of type 'Iterable<string> | LazySet<string>'.
```

File: lib/serialization/ObjectMiddleware.js (hunk 1 above)
Comment (jakebailey) [reply lib/serialization/ObjectMiddleware.js:55]: Just to be clear, you're saying that before this change, uses of this method were not caught, and that's the bug? Just trying to understand what is is to be testing.

File: lib/util/LazySet.js (hunk 2 above)
Comment (jakebailey) [reply lib/util/LazySet.js:201]: I was able to minimize it, but, it's not new to TS 5.0 it seems.

Comment (alexander-akait) [reply lib/util/LazySet.js:201]: @jakebailey Can you provide a link to the issue?

Comment (jakebailey) [reply lib/util/LazySet.js:201]: Just filed: https://github.com/microsoft/TypeScript/issues/53967

File: lib/serialization/ObjectMiddleware.js (hunk 1 above)
Comment (alexander-akait) [reply lib/serialization/ObjectMiddleware.js:55]: @jakebailey Yeah, this is not a bug, but rather an improvement, just reported that this problem was not caught before

Comment (jakebailey) [reply lib/serialization/ObjectMiddleware.js:55]: Ah, gotcha.

**Review by TheLarkInn — APPROVED** [review]

**Comment by TheLarkInn** [comment]
Really nice work @alexander-akait. We got some more helpful types out of it too. Time to get type coverage up to 80% now 😈

**Pull request merged**